File: MoireStudies/PatternView.m
//
//  PatternView.m
//  MoireStudies
//
//  Created by Jialiang Xiang on 2020-12-08.
//

#import "PatternView.h"
#import "ControlView.h"
#import "SliderControlView.h"

@implementation PatternView {
    UIView* mask;
    ControlView* controlView;
}

- (PatternView*)initWithFrame:(CGRect)f ControlFrame:(CGRect)c {
    self = [super init];
    if (self) {
        self.backgroundColor = [UIColor yellowColor];
        self.frame = f;
        [self setUpControlsWithControlFrame:c];
    }
    return self;
}

- (void)setUpControlsWithControlFrame: (CGRect)c {
    controlView = [[SliderControlView alloc]initWithFrame:c];
    [self addSubview:controlView];
}

@end

File: MoireStudies/PatternView.swift
//
//  PatternView.swift
//  MoireStudies
//
//  Created by Jialiang Xiang on 2020-12-23.
//

import Foundation
import UIKit

class PatternView : UIView {
    private var _pattern: Pattern?
    var pattern: Pattern {
        get {
            if let p = _pattern {
                return p
            } else {
                return Pattern.defaultPattern()
            }
        }
        set {
            if _pattern != newValue {
                self.patternChanged(from: _pattern, to: newValue)
            }
        }
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
    }
    
    func setUp(pattern: Pattern) {
        preconditionFailure("this method must be overridden")
    }
    
    func patternChanged(from oldPattern: Pattern?, to newPattern: Pattern) {
        preconditionFailure("this method must be overridden")
    }
}
